python/tvm/contrib/pipeline_executor.py

```diff
@@ -0,0 +1,534 @@
+                    # Do acyclic check after increase the in-degree.
+                    if not self.check_dag_acyclic(
+                        binding.io_owner, self.io_owner.input_bindings.bindings
+                    ):
+                        raise RuntimeError(f"Illegal connection: cause a circle!")
+
+        def __init__(self, mod=None):
+            """init class"""
+            self.target_host = None
+            self.build_func = None
+            self.params = None
+            self.target = None
+            self.name = None
+            self.dev = None
+            self.idx = None
+            self.mod = mod
+            self.input_params = InferType()(mod)["main"].params
+            self.output_values = InferType()(mod)["main"].checked_type.ret_type
+            self.input_bindings = PipelineConfig.BindingList(self, "input")
+            self.output_bindings = PipelineConfig.BindingList(self, "output")
+
+        def __eq__(self, other):
+            """check if self equl other"""
+            if isinstance(other, PipelineConfig.ModuleWrapper):
+                return self.mod == other.mod
+
+            return False
+
+        def __getitem__(self, key):
+            """get item by key"""
+            if isinstance(key, str):
+                if key == "input":
+                    return self.input_bindings
+
+                if key == "output":
+                    return self.output_bindings
+
+            raise RuntimeError(f"{key} not found!")
+
+        def get_data_type(self, key, stype):
+            """get module input/output data type."""
+            if stype == "input":
+                for param in self.input_params:
+                    if param.name_hint == key:
+                        return param._checked_type_
+
+            if stype == "output":
+                if isinstance(self.output_values, tvm.ir.type.TupleType):
+                    if int(key) < len(self.output_values.fields):
+                        return self.output_values.fields[int(key)]
+                elif int(key) == 0:
+                    return self.output_values
+
+            return None
+
+        def set_idx_name(self, idx):
+            """generate name by idx and storage idx value"""
+            self.idx = idx
+            self.name = "mod{}".format(str(idx))
+
+        def is_root_mod(self):
+            """use by dag topology sort, identify if this item is root item"""
+            return all([not b.parents for b in self.input_bindings.bindings.values()])
+
+        def remove_self_from_bindings(self):
+            """use by DAG topology sort, by remove self from binding to reduce child in-degree"""
+            for binding in self.output_bindings.bindings.values():
+                for child in binding.bindings:
+                    if binding in child.parents:
+                        child.parents.remove(binding)
+
+    class BindingList:
+        """Use to storage Binding list.
```

Comment:
This doesn't make sense.

Reply:
changed into "Container for bindings(input or output interface)"